qontract_api/qontract_api/integrations/slack_usergroups/slack_workspace_client.py

```diff
@@ -0,0 +1,484 @@
+                        for pk in ug.prefs.channels
+                        if pk in channel_name_by_id
+                    ],
+                ),
+            )
+            for ug in usergroups
+        ]
+
+    def clean_slack_usergroups(
+        self, usergroups: Iterable[SlackUsergroup]
+    ) -> list[SlackUsergroup]:
+        """Clean usergroups by removing non-existing users/channels (efficient batch operation).
+
+        Args:
+            usergroups: List of usergroups to clean
+
+        Returns:
+            List of cleaned usergroups with only valid users/channels
+        """
+        # Pre-fetch all data ONCE for all usergroups (2 cache lookups total!)
+        # Build O(1) lookup sets ONCE
+        valid_org_names = {
+            user.org_username for user in self.get_users().values() if not user.deleted
+        }
+        valid_channel_names = {channel.name for channel in self.get_channels().values()}
+
+        # Filter all usergroups in-memory (very fast!)
+        return [
+            SlackUsergroup(
+                handle=ug.handle,
+                config=SlackUsergroupConfig(
+                    description=ug.config.description,
+                    users=[name for name in ug.config.users if name in valid_org_names],
+                    channels=[
+                        name
+                        for name in ug.config.channels
+                        if name in valid_channel_names
+                    ],
+                ),
+            )
+            for ug in usergroups
+        ]
+
+    def create_usergroup(
+        self,
+        *,
+        handle: str,
+        name: str | None = None,
+    ) -> SlackUsergroupAPI:
+        """Create a new usergroup and add to cache.
+
+        Args:
+            handle: Usergroup handle (e.g., "oncall-team")
+            name: Usergroup display name (defaults to handle)
+
+        Returns:
+            Created SlackUsergroupAPI object
+        """
+        created_ug = self.slack_api.usergroup_create(handle=handle, name=name)
+
+        # Add to cache
+        try:
+            with self.cache.lock(self._cache_key_usergroups()):
+                cached = self._get_cached_usergroups(self._cache_key_usergroups())
+                if cached:
+                    cached[created_ug.id] = created_ug
+                    self._set_cached_usergroups(
+                        self._cache_key_usergroups(),
+                        cached,
+                        self.settings.slack.usergroup_cache_ttl,
+                    )
+        except RuntimeError as e:
+            logger.warning(f"Could not acquire lock to cache created usergroup: {e}")
+
+        return created_ug
+
+    def update_usergroup(
+        self,
+        *,
+        handle: str,
+        name: str | None = None,
+        description: str | None = None,
+        channels: Iterable[str] | None = None,
+    ) -> SlackUsergroupAPI:
+        """Update usergroup and cache (O(1) update, not invalidation).
```

Comment:
but this method will be called per update action, can we aggregate all updates in one call to only update cache once?